File: dependency-inversion-ios/DependencyInversion/CachingFruitFetcher.swift
import Foundation

class CachingFruitFetcher {
    
    func fetchFruit() -> Fruit? {
        if let name = UserDefaults.standard.string(forKey: "lastFruit") {
            return Fruit(name: name)
        } else {
            let maybeFruit = try? InternetFruitRepository.instance.loadFruitFromInternet()
            UserDefaults.standard.set(maybeFruit?.name, forKey: "lastFruit")
            return maybeFruit
        }
    }
    
}

File: dependency-inversion/android/default/lib/src/main/java/com/example/CachingFruitFetcher.java
package com.example;

public class CachingFruitFetcher {

    Fruit fetchFruit() {
        if (Preferences.instance.valueForKey("lastFruit") != null) {
            return new Fruit(Preferences.instance.valueForKey("lastFruit"));
        } else {
            Fruit fruit = InternetFruitRepository.instance.loadFruitFromInternet();
            Preferences.instance.put("lastFruit", fruit.getName());
            return fruit;
        }
    }

}
